### ISSUE TITLE
Unsound error while check if object instance of another

### ISSUE THREAD

**Issue by Crinax** [microsoft/TypeScript#56143]
### 🔎 Search Terms

unsound, instanceof, instance, object, types, error

### 🕗 Version & Regression Information

- This is unsound error
- All versions in playground have this bug

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/MYewdgziA2CmB00QHMAUAVAngB1gUQCcCQCACAS0gBcBDMYWEAM1NVQEpSBeAPlIGZ27IA

### 💻 Code

```ts
console.log(TypeError instanceof (() => 3));
```

### 🙁 Actual behavior

TypeScript successfully transpiles the code

### 🙂 Expected behavior

Throws error because the right-side operand has not prototype

**Comment by MartinJohns** [comment]
The right side of `instanceof` must just be of the type `any` or assignable to the `Function` interface. Having a prototype is no requirement. `() => 3` is assignable to `Function`.

**Comment by Crinax** [comment]
> The right side of `instanceof` must just be of the type `any` or assignable to the `Function` interface. Having a prototype is no requirement. `() => 3` is assignable to `Function`.

Right side must be of Constructor type, not a Function, since not all functions have a prototype. You can see it in the MDN documentation: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof

**Comment by MartinJohns** [comment]
> Right side must be of Constructor type, not a Function,

I was referring to the TypeScript side, which has a clear error message when you provide another type (e.g. `{}`). So the current behavior is working as intended and not a bug. You might want to rephrase it to a feature request if you think this should be changed.



> You can see it in the MDN documentation: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof

The documentation of this site provides examples where you're not dealing with constructors and just plain functions.

```javascript
function C() {}

const o = new C();

// true, because: Object.getPrototypeOf(o) === C.prototype
o instanceof C;
```

**Comment by fatcerberus** [comment]
To expand on Martin's comment: basically TS makes no distinction between `function f(): T {}` (which *does* have a prototype and might actually be a constructor) and `() => T` (which doesn't).  Both are the exact same type to the compiler.  So that's why the `instanceof` check is allowed.

**Comment by JoshuaKGoldberg** [comment]
https://github.com/microsoft/TypeScript/pull/56197#issuecomment-1777656189:

> I'm starting to doubt why I didn't just Won't Fix the linked issue. Who would ever do this? It seems like an impossible mistake to make.

I can see an argument that when refactoring an area of code from classes to non-class functions. #56197 should handle this kind of case:

```ts
const ThisUsedToBeAClass = () => {};

TypeError instanceof ThisUsedToBeAClass;
```

...though I've never seen this in the wild.

**Comment by RyanCavanaugh** [comment]
I guess the problem is that #56197 *only* fixes this when we can see the value declaration of the expression. For cases like
```ts
const obj1: () => void = function f() { }
const obj2: () => void = () => { }

{} instanceof obj1;
{} instanceof obj2;
```
There's nothing distinguishing the bottom two lines from the type system's perspective

**Comment by JoshuaKGoldberg** [comment]
Perhaps it'd make sense to add an understanding of that to the system? Help it differentiate between arrow functions and the other kinds?

**Comment by RyanCavanaugh** [comment]
I think it's a hard sell, since it creates the possible mistake of people accidently *requiring* a function-with-a-prototype when they don't actually care (which is by far the common case).

**Comment by JoshuaKGoldberg** [comment]
Hmm, yeah. I started going through the process of filing an issue asking that the info be stored on the `ts.Type`. But even more so than `{ (): ... }` vs `() => ...`, +1 that the possible mistake probability is high... ah well.
